### PR TITLE
Replace boost::filesystem with std::filesystem

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Signed-off-by: Zackery Spytz <zspytz@gmail.com>

### CODE REVIEW

**Comment by mormj** [comment]
@ZackerySpytz - Thank you for the boost removal!  The only issue here is going to be compiler support for C++17 features.  std::filesystem appears to only be supported (https://en.cppreference.com/w/cpp/compiler_support) in 
- gcc-8
- MSVC 19.4 (VS 2017)
- clang 7
- apple-clang 11
So we need a way to ensure that building from source is easily possible across the major platforms without jumping through hoops.  Need to gather more info here as to what the range of support is / should be.

**Comment by mbr0wn** [comment]
I like the change, but the linker errors on Debian and CentOS need to be resolved somehow.

**Comment by skoslowski** [comment]
This should fix gcc8 builds (CentOS and Debian). 
```cmake
target_link_libraries(gnuradio-runtime PUBLIC $<$<AND:$<CXX_COMPILER_ID:GNU>,$<VERSION_LESS:$<CXX_COMPILER_VERSION>,9.0>>:stdc++fs>)
```
FEC and FFT should also link to that lib (but with PRIVAT)

(see https://gcc.gnu.org/gcc-9/changes.html --> Runtime Library)

**Review by mbr0wn — CHANGES_REQUESTED** [review]
Please fix the linker errors (see the comment by @skoslowski). Thanks!

File: gnuradio-runtime/lib/constants.cc.in [hunk 1]
```diff
@@ -26,28 +26,28 @@ const std::string prefix()
     if (prefix != NULL)
         return prefix;
 
-    boost::filesystem::path prefix_rel_lib = "@prefix_relative_to_lib@";
-    boost::filesystem::path gr_runtime_lib_path = boost::dll::this_line_location();
+    std::filesystem::path prefix_rel_lib = "@prefix_relative_to_lib@";
+    std::filesystem::path gr_runtime_lib_path = boost::dll::this_line_location().string();
```
Comment (ryanvolz) [inline gnuradio-runtime/lib/constants.cc.in:30]:
```suggestion
    std::filesystem::path gr_runtime_lib_path = boost::dll::this_line_location();
```
Adding the .string() is not necessary with the CMake fix (discussed below).

**Review by ryanvolz — CHANGES_REQUESTED** [review]
I build this successfully with Visual Studio 2017 after the following addition:

https://github.com/ryanvolz/gnuradio/commit/9d830714e786573c06ae7940d3d1f8d1e3c2050a

Basically, the `boost::dll` that we use in `constants.cc` pulls in `boost::filesystem` still unless you define `BOOST_DLL_USE_STD_FS`. The CMake change in the above commit adds that definition and reverts a prior change that evidently worked round the needed `boost::filesystem::path` to `std::filesystem::path` conversion.

I also tried my hand at the conditional linking for GCC < 9.0:

https://github.com/ryanvolz/gnuradio/commit/15f348bfb0f1fe0d705a589f35cf37111ad9a951

Both of those are in my branch: https://github.com/ryanvolz/gnuradio/tree/boost-std-filesystem.

**Comment by mormj** [comment]
So this probably would only impact MSVC (with the additional changes) - currently our MSVC min version is
```
set(MSVC_MIN_VERSION "1900")
```
(VS 2015) and would need to be bumped - presumably to `1904` (VS 2017).  Not being a Windows user/developer, will need some feedback on the impact of this. @ryanvolz

**Comment by mormj** [comment]
@ZackerySpytz - will you have a chance to update the PR with the suggested changes?

**Comment by mormj** [comment]
pushing forward in https://github.com/gnuradio/gnuradio/pull/4368 since this has gone quiet.  Thank you @ZackerySpytz for the contributions - your commits are included in that PR.

**Pull request closed without merging**